### ISSUE TITLE
Expose parameter expect_string in PaloAltoPanosBase commit() method

### ISSUE THREAD

**Issue by hkam40k** [ktbyers/netmiko#3033]
Hello,

Currently the `commit` method has a hardcoded `expect_string="100%"` when calling `self.send_command_expect()`. If the response of the commit is anything else than a full success, the commit will fail with an error `Search pattern never detected in send_command: 100%".

It would be good to expose the `expect_string` parameter in the `commit` method parameters, so that users can customize the search parameter to account for common errors, such as validation errors and lock errors.

The following method signature for `commit` would catch validation errors, commit-lock errors (somebody else holds the commit-lock), and also successful commits:
```
def commit(
        self,
        comment=None,
        force=False,
        partial=False,
        device_and_network=False,
        policy_and_objects=False,
        vsys="",
        no_vsys=False,
        expect_string="(Validation [Ee]rror|Server [Ee]rror|100%)",
        delay_factor=0.1,
    ):
```
(Unfortunately, at least some PA platforms, seem to return the string "error" in lower or upper case depending on type of error, so it does not hurt to be prepared in case PA makes it more uniform one day...)

If you agree with the above, I can submit a PR. Even if the default value for `expect_string` is kept `"100%"` to be more backwards compatible, at least exposing the parameter would still be a good idea.

**Comment by ktbyers** [comment]
We could just make the regex search not care about capitalization.

What about just changing the one existing line to:

```python
expect_string=r"error|100%"
```

Formatted correctly so that capitalization doesn't matter i.e. re.I flag (essentially).

I am not very keen on adding an argument to the method for this (but let me know if I am missing something).

**Comment by hkam40k** [comment]
> What about just changing the one existing line to:
> 
> ```python
> expect_string=r"error|100%"
> ```
> 
> Formatted correctly so that capitalization doesn't matter i.e. re.I flag (essentially).

While catching the string "error" would maybe catch more types of errors that I can't currently think of, it may be too generic, since currently it would also detect a commit with a message with the word "error" in it to be a hit, resulting in prematurely exiting the commit progress polling, and eventually raising an error when `commit_marker` is not detected in the output.

Would a compromise of `expect_string=r"validation error|server error|100%"` be slightly more safe to use? Still not completely bulletproof but I guess it's not possible to get around it since the `command_string` gets echoed from the channel.

> I am not very keen on adding an argument to the method for this (but let me know if I am missing something).

I think it would be a useful way to handle other type of unknown cases if they arise unexpectedly (different types of errors, possible changes in the error messages themselves in future updates or different models etc), but otherwise I also do not see a use for it.

**Comment by ktbyers** [comment]
The difficulty/problem with potentially adding it as an argument arises when Netmiko gets integrated in other libraries and they just call `commit()` (i.e. the user of these other libraries don't have an easy way of passing in arguments).

The second difficulty is I would really like to integrate the `commit()` method in Netmiko to use the same set of arguments for all platforms (though this will definitely take an overhaul as they are all over the place right now)...so I am reluctant to add a new argument that will make this harder.

This is fine with me:

```
Would a compromise of expect_string=r"validation error|server error|100%" 
```

I assume the "error" case implies the commit did not go through (and we should consequently raise an exception)?

**Comment by hkam40k** [comment]
> The difficulty/problem with potentially adding it as an argument arises when Netmiko gets integrated in other libraries and they just call `commit()` (i.e. the user of these other libraries don't have an easy way of passing in arguments).

This is true. But at least there would be an option for libraries to pass in the argument (or for users to override and then pass the argument to get around it). I would say that a library that does not permit the user to utilize the optional arguments for the commit method is doing it wrong already, since there are other useful optional args there such as comment and delay_factor. The suggested change for the `expect_string` value already does help a lot though.

> The second difficulty is I would really like to integrate the `commit()` method in Netmiko to use the same set of arguments for all platforms (though this will definitely take an overhaul as they are all over the place right now)...so I am reluctant to add a new argument that will make this harder.

Understandable. I'm sure this is a constant challenge.

> This is fine with me:
> 
> ```
> Would a compromise of expect_string=r"validation error|server error|100%" 
> ```
> 
> I assume the "error" case implies the commit did not go through (and we should consequently raise an exception)?

What I mean here is for example with a commit command `commit description "Change abc to fix error xyz with..."`, netmiko would stop polling when encountering the string "error" and raise an exception to the caller, even if the commit actually does succeed. Therefore the above suggestion of `expect_string=r"validation error|server error|100%"` to make the expected outputs slightly more specific to not trip on the string "error" alone. I'm not sure if it can be avoided that the input `command_string` gets parsed from the channel when reading the output of the SSH command.

So going with `expect_string=r"validation error|server error|100%" ` seems good to me.

**Comment by ktbyers** [comment]
```
So going with expect_string=r"validation error|server error|100%" seems good to me.
```

That works for me. Do you want to do a PR on this?

**Comment by hkam40k** [comment]
> That works for me. Do you want to do a PR on this?

I can send a PR, yes.

Do you think it would be feasible to do a release for the previous 3.x version also, i.e. a patch version 3.4.1 with the discussed change in place? I think the related code is similar (or same) in both versions, so the same commit could potentially be merged into both the latest version, and also to 3.4.0.

**Comment by ktbyers** [comment]
Nope, Netmiko 3.x is dead. You would have to maintain your own fork for that.

**Comment by hkam40k** [comment]
> Nope, Netmiko 3.x is dead. You would have to maintain your own fork for that.

Ok, well, that shouldn't be a big problem anyway.

**Comment by hkam40k** [comment]
FYI, I've identifier another possible error case during commits, with a different error message:
```
admin@PA-VM# commit
Commit job 225 is in progress. Use Ctrl+C to return to command prompt
..........55%.
Error: application 'ntp-base' not found
(Module: device)
Commit failed
```
So I would add `commit failed` to expected strings:
```
expect_string=r"validation error|server error|commit failed|100%"
```
The word `error` of course could be a common denominator, but I still would not do it for the reasons previously mentioned in the discussion. It is possible that there are other error types that will not get caught with these, but it is still an improvement compared to previous.

**Comment by ktbyers** [comment]
@hkam40k Yes, `commit failed` looks reasonably/good to me.
